File: Framework/CSharp/Framework/Framework/SmartMath.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Windows;
using System.Xml.XPath;

namespace Smartkernel.Framework
{
	/// <summary>
	/// 数学辅助类
	/// </summary>
	public static class SmartMath
	{
		/// <summary>
		/// 求质因数
		/// </summary>
		/// <typeparam name="T">数字类型</typeparam>
		/// <param name="number">待求质因数的数字，必须大于0。且是整数</param>
		/// <returns>质因数列表</returns>
		public static List<T> GetPrimeDivisors<T>(T number) where T : IComparable
		{
			var list = new List<T>();

			for (var i = SmartGeneric.ToGeneral<T>(2); SmartGeneric.LessThanOrEquals(i, number); i = SmartGeneric.Increase(i))
			{
				while (SmartGeneric.NotEquals(number, i))
				{
					var divideRemainder = SmartGeneric.DivideRemainder(number, i);
					if (SmartGeneric.Equals(divideRemainder, SmartGeneric.ToGeneral<T>(0)))
					{
						list.Add(i);
						number = SmartGeneric.Divide(number, i);
					}
					else
					{
						break;
					}
				}
			}
			list.Add(number);

			return list;
		}

		/// <summary>
		/// 求最大公约数，如果没有，则返回1
		/// </summary>
		/// <typeparam name="T">数字类型</typeparam>
		/// <param name="numbers">待求的数字列表</param>
		/// <returns>最大公约数</returns>
		public static T GetMaxDivisor<T>(IEnumerable<T> numbers) where T : IComparable
		{
			var numberList = new List<T>(numbers);
			var commonPrimeDivisors = new List<KeyValuePair<T, int>>();
			bool isFirst = true;
			numberList.ForEach(number =>
				{
					var list = new List<KeyValuePair<T, int>>();
					GetPrimeDivisors(number).ForEach(primeDivisor =>
						{
							int i = 0;
							while (list.Contains(new KeyValuePair<T, int>(primeDivisor, i)))
							{
								i++;
							}
							list.Add(new KeyValuePair<T, int>(primeDivisor, i));
						});
					if (isFirst)
					{
						isFirst = false;
						commonPrimeDivisors.AddRange(list);
					}
					else
					{
						commonPrimeDivisors = commonPrimeDivisors.Intersect(list).ToList();
					}
				});

			//所有公共质因数的乘积就是最大公约数
			var result = SmartGeneric.ToGeneral<T>(1);
			commonPrimeDivisors.ForEach(item =>
				{
					result = SmartGeneric.Multiply(result, item.Key);
				});
			return result;
		}

		/// <summary>
		/// 二进制转换为十进制
		/// </summary>
		/// <param name="input">待转换的字符串</param>
		/// <returns>转换的结果</returns>
		public static int BinaryToDecimal(string input)
		{
			return Convert.ToInt32(input, 2);
		}

		/// <summary>
		/// 八进制转换为十进制
		/// </summary>
		/// <param name="input">待转换的字符串</param>
		/// <returns>转换的结果</returns>
		public static int OctalToDecimal(string input)
		{
			return Convert.ToInt32(input, 8);
		}

		/// <summary>
		/// 十六进制转换为十进制
		/// </summary>
		/// <param name="input">待转换的字符串</param>
		/// <returns>转换的结果</returns>
		public static int HexToDecimal(string input)
		{
			return Convert.ToInt32(input, 16);
		}

		/// <summary>
		/// 十进制转换为二进制
		/// </summary>
		/// <param name="input">待转换的数字</param>
		/// <returns>转换的结果</returns>
		public static string DecimalToBinary(int input)
		{
			return Convert.ToString(input, 2);
		}

		/// <summary>
		/// 十进制转换为八进制
		/// </summary>
		/// <param name="input">待转换的数字</param>
		/// <returns>转换的结果</returns>
		public static string DecimalToOctal(int input)
		{
			return Convert.ToString(input, 8);
		}

		/// <summary>
		/// 十进制转换为十六进制
		/// </summary>
		/// <param name="input">待转换的数字</param>
		/// <returns>转换的结果</returns>
		public static string DecimalToHex(int input)
		{
			return Convert.ToString(input, 16);
		}

		/// <summary>
		/// 角度转换为弧度
		/// </summary>
		/// <param name="input">待转换的角度值</param>
		/// <returns>弧度表示</returns>
		public static decimal AngleToRadian(decimal input)
		{
			return (Convert.ToDecimal(Math.PI) * input) / 180m;
		}

		/// <summary>
		/// 弧度转换为角度
		/// </summary>
		/// <param name="input">待转换的弧度值</param>
		/// <returns>角度表示</returns>
		public static decimal RadianToAngle(decimal input)
		{
			return input * 180m / Convert.ToDecimal(Math.PI);
		}

		/// <summary>
		/// 角度转换为弧度
		/// </summary>
		/// <param name="input">待转换的角度值</param>
		/// <returns>弧度表示</returns>
		public static double AngleToRadian(double input)
		{
			return (Math.PI * input) / 180d;
		}

		/// <summary>
		/// 弧度转换为角度
		/// </summary>
		/// <param name="input">待转换的弧度值</param>
		/// <returns>角度表示</returns>
		public static double RadianToAngle(double input)
		{
			return input * 180d / Math.PI;
		}

		/// <summary>
		/// 计算利息
		/// </summary>
		/// <param name="principalMoney">本金</param>
		/// <param name="year">年数</param>
		/// <param name="rate">年利率</param>
		/// <param name="isCompoundInterest">是否复利</param>
		/// <returns>本利和</returns>
		public static double CalculateInterest(double principalMoney, double year, double rate, bool isCompoundInterest = true)
		{
			if (isCompoundInterest)
			{
				return Math.Pow(1 + rate, year) * principalMoney;
			}
			return (1 + rate * year) * principalMoney;
		}

		/// <summary>
		/// 计算数学表达式的值（只支持加减乘除）
		/// </summary>
		/// <param name="expression">数学表达式</param>
		/// <returns>结果</returns>
		public static double Eval(string expression)
		{
			var xpath = string.Format("number({0})", new Regex(@"([\+\-\*])").Replace(expression, " ${1} ").Replace("/", " div ").Replace("%", " mod "));

			return (double)new XPathDocument(new System.IO.StringReader("<r/>")).CreateNavigator().Evaluate(xpath);
		}

		/// <summary>
		/// 通过角度计算偏移
		/// </summary>
		/// <param name="angle">角度</param>
		/// <param name="distance">距离</param>
		/// <returns>结果</returns>
		public static Point GetOffsetFromAngleAndDistance(double angle, double distance)
		{
			double x = Math.Cos(AngleToRadian(angle)) * distance;
			double y = Math.Tan(AngleToRadian(angle)) * x;
			return new Point(x, y);
		}

		/// <summary>
		/// 通过偏移计算角度
		/// </summary>
		/// <param name="offset">偏移</param>
		/// <returns>结果</returns>
		public static double GetAngleFromOffset(Point offset)
		{
			double opposite = Math.Abs(offset.Y);
			double adjacent = Math.Abs(offset.X);

			if (offset.Y < 0)
			{
				opposite = -opposite;
			}

			double angle = RadianToAngle(Math.Atan(opposite / adjacent));
			if (double.IsNaN(angle))
			{
				return 0.0;
			}

			if (offset.X < 0)
			{
				angle = 90 + (90 - angle);
			}

			return angle;
		}
	}
}

File: Framework/Php/framework/framework/smartmath.php
<?php

/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */

namespace Smartkernel\Framework {

    use Smartkernel\Framework\SmartGeneric;
    use Smartkernel\Framework\Collections\SmartKeyValuePair;

/*
     * 数学辅助类
     */

    class SmartMath {
        /*
         * 取绝对值
         * 
         * @param number $input，输入
         * @return float，结果
         * 
         */

        public static function abs($input) {
            return abs($input);
        }

        /*
         * 返回大于等于输入值的最小整数
         * 
         * @param number $input，输入
         * @return float，结果
         * 
         */

        public static function ceiling($input) {
            return ceil($input);
        }

        /*
         * 返回小于等于输入值的最大整数
         * 
         * @param number $input，输入
         * @return float，结果
         * 
         */

        public static function floor($input) {
            return floor($input);
        }

        /*
         * 取最大值
         * 
         * @param number $left，输入
         * @param number $right，输入
         * @return float，结果
         * 
         */

        public static function max($left, $right) {
            return max($left, $right);
        }

        /*
         * 取最小值
         * 
         * @param number $left，输入
         * @param number $right，输入
         * @return float，结果
         * 
         */

        public static function min($left, $right) {
            return min($left, $right);
        }

        /*
         * 计算幂
         * 
         * @param number $x，输入
         * @param number $y，幂
         * @return float，结果
         * 
         */

        public static function pow($x, $y) {
            return pow($x, $y);
        }

        /*
         * 按指定的小数位舍入
         * 
         * @param number $d，输入
         * @param int $decimals，小数位数
         * @return float，结果
         * 
         */

        public static function round($d, $decimals) {
            return round($d, $decimals);
        }

        /*
         * 获取符号（-1为负数，1为正数）
         * 
         * @param number $input，输入
         * @return int，结果
         * 
         */

        public static function sign($input) {
            if ($input > 0) {
                return 1;
            } else if ($input < 0) {
                return -1;
            } else {
                return 0;
            }
        }

        /*
         * 开平方
         * 
         * @param number $input，输入
         * @return float，结果
         * 
         */

        public static function sqrt($input) {
            return sqrt($input);
        }

        /*
         * 计算表达式
         * 
         * @param string $expression，表达式
         * @return float，结果
         * 
         * */

        public static function evalExpression($expression) {
            eval('$result = ' . $expression . ';');

            return $result;
        }

        /*
         * 求质因数
         * 
         * @param int $number，待求质因数的数字，必须大于0。且是整数
         * @return array，质因数列表
         * 
         */

        public static function getPrimeDivisors($number) {
            $number = intval($number);
            $list = array();

            for ($i = 2; SmartGeneric::lessThanOrEquals($i, $number); $i = SmartGeneric::increase($i)) {

                while (SmartGeneric::notEquals($number, $i)) {
                    $divideRemainder = SmartGeneric::divideRemainder($number, $i);
                    if (SmartGeneric::equals($divideRemainder, 0)) {
                        $list[] = $i;
                        $number = SmartGeneric::divide($number, $i);
                    } else {
                        break;
                    }
                }
            }
            $list[] = $number;

            return $list;
        }

        /*
         * 求最大公约数，如果没有，则返回1
         * 
         * $param array $numbers，待求的数字列表
         * $return int，最大公约数
         * 
         */

        public static function getMaxDivisor(array $numbers) {
            $numberList = $numbers;
            $commonPrimeDivisors = array();

            foreach ($numberList as $number) {
                $list = array();

                $primeDivisors = self::getPrimeDivisors($number);

                foreach ($primeDivisors as $primeDivisor) {
                    $i = 0;
                    while (SmartArray::contains($list, new SmartKeyValuePair($primeDivisor, $i))) {
                        $i++;
                    }
                    $list[] = new SmartKeyValuePair($primeDivisor, $i);
                }

                if (count($commonPrimeDivisors) == 0) {
                    $commonPrimeDivisors = SmartArray::addRange($commonPrimeDivisors, $list);
                } else {
                    $commonPrimeDivisors = SmartArray::intersect($commonPrimeDivisors, $list);
                }
            }

            //所有公共质因数的乘积就是最大公约数
            $result = 1;

            foreach ($commonPrimeDivisors as $item) {
                $result = SmartGeneric::multiply($result, $item->Key);
            }
            return $result;
        }

        /*
         * 二进制转换为十进制
         * 
         * @param string $input，待转换的字符串
         * @return float，转换的结果
         * 
         */

        public static function binaryToDecimal($input) {
            return base_convert($input, 2, 10);
        }

        /*
         * 八进制转换为十进制
         * 
         * @param string $input，待转换的字符串
         * @return float，转换的结果
         * 
         */

        public static function octalToDecimal($input) {
            return base_convert($input, 8, 10);
        }

        /*
         * 十六进制转换为十进制
         * 
         * @param string $input，待转换的字符串
         * @return float，转换的结果
         * 
         */

        public static function hexToDecimal($input) {
            return base_convert($input, 16, 10);
        }

        /*
         * 十进制转换为二进制
         * 
         * @param string $input，待转换的字符串
         * @return float，转换的结果
         * 
         */

        public static function decimalToBinary($input) {
            return base_convert($input, 10, 2);
        }

        /*
         * 十进制转换为八进制
         * 
         * @param string $input，待转换的字符串
         * @return float，转换的结果
         * 
         */

        public static function decimalToOctal($input) {
            return base_convert($input, 10, 8);
        }

        /*
         * 十进制转换为八进制
         * 
         * @param string $input，待转换的字符串
         * @return float，转换的结果
         * 
         */

        public static function decimalToHex($input) {
            return base_convert($input, 10, 16);
        }

        /*
         * 角度转换为弧度
         * 
         * @param float $input，待转换的角度值
         * @return float，弧度表示
         * 
         */

        public static function angleToRadian($input) {
            return (M_PI * $input) / 180;
        }

        /*
         * 弧度转换为角度
         * 
         * @param float $input，待转换的角度值
         * @return float，角度表示
         * 
         */

        public static function radianToAngle($input) {
            return $input * 180 / M_PI;
        }

        /*
         * 计算利息
         * 
         * @param float $principalMoney，本金
         * @param int $year，年数
         * @param float $rate，年利率
         * @param bool $isCompoundInterest，是否复利
         * @return float，本利和
         * 
         */

        public static function calculateInterest($principalMoney, $year, $rate, $isCompoundInterest = true) {
            if ($isCompoundInterest) {
                return self::pow(1 + $rate, $year) * $principalMoney;
            }
            return (1 + $rate * $year) * $principalMoney;
        }

        /*
         * 通过角度计算偏移
         * 
         * @param float $angle，角度
         * @param float $distance，距离
         * @return point，结果
         * 
         */

        public static function getOffsetFromAngleAndDistance($angle, $distance) {
            $x = cos(self::angleToRadian($angle)) * $distance;
            $y = tan(self::angleToRadian($angle)) * $x;
            return array("x" => $x, "y" => $y);
        }

        /*
         * 通过偏移计算角度
         * 
         * @param point $offset，偏移
         * @param float，结果
         * 
         */

        public static function getAngleFromOffset(array $offset) {
            $opposite = self::abs($offset["y"]);
            $adjacent = self::abs($offset["x"]);

            if ($offset["x"] < 0) {
                $opposite = -$opposite;
            }

            $angle = self::radianToAngle(atan($opposite / $adjacent));

            if ($offset["x"] < 0) {
                $angle = 90 + (90 - $angle);
            }

            return $angle;
        }

    }

}